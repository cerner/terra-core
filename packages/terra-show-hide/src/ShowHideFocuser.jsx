import React, { forwardRef, useState } from 'react';
import PropTypes from 'prop-types';
import classNamesBind from 'classnames/bind';
import { v4 as uuidv4 } from 'uuid';
import styles from './ShowHide.module.scss';

const cx = classNamesBind.bind(styles);

const propTypes = {
  /**
   * The portion of the text that gets focus once the hidden portion is disclosed to be announced by assistive technologies
   */
  focusableText: PropTypes.string.isRequired,
  /**
   * Text that precedes the focusable portion.
   */
  prefix: PropTypes.string,
};

/**
 * @param string - string value
 * @returns a trimmed string with trailing white space added
 */
const addTrailingWhiteSpace = (string) => `${string.trim()} `;

/**
 * @param string - string value
 * @returns a trimmed string with preceding whitespace added
 */
const addHeadingWhiteSpace = (string) => ` ${string.trim()}`;

/**
 * This method splits the string in two portions without splitting words. The first (focusable) portion's length should not exceed max length for assistive technologies in the text mode being able to use Arrow Up?Down correctly.
 * @param string - the text
 * @returns the array of strings, where the first value is short focusable string and the second value is the rest of the string (if such exists)
 */
const truncate = (string) => {
  if (!string) {
    return null;
  }
  const text = string.trim();
  const maxLength = 155;
  if (text.length < maxLength) {
    return [text];
  }
  const lastFullWordEndPosition = text.slice(0, maxLength).lastIndexOf(' ');
  const first = text.slice(0, lastFullWordEndPosition)?.trim();
  const second = text.slice(lastFullWordEndPosition)?.trim();
  return [first, second];
};

/**
 * This component splits the text with dividing spans to allow the assistive technologies focus on the part of the text
 */
const ShowHideFocuser = forwardRef((props, ref) => {
  const {
    prefix,
    focusableText,
  } = props;

  const truncatedText = truncate(focusableText);
  const [focusable, setFocusable] = useState(true);

  const id = uuidv4();
  const refId = `focus-ref-${id}`;

  const onBlur = () => setFocusable(false);

  return (
    <>
      {prefix ? addTrailingWhiteSpace(prefix) : null}
      {/* this divider starts a new 'line' for JAWS when reading line-by-line */}
      <span className={cx('text-divider')} />
      <span
        id={refId}
        ref={ref}
        tabIndex={focusable ? '-1' : null}
        role={focusable ? 'group' : null}
        onBlur={onBlur}
      >
        {truncatedText[0]}
      </span>
      {/* this divider ends a line for JAWS when reading line-by-line */}
      <span className={cx('text-divider')} />
      {truncatedText[1] ? addHeadingWhiteSpace(truncatedText[1]) : null}
    </>
  );
});

ShowHideFocuser.propTypes = propTypes;
export default ShowHideFocuser;
