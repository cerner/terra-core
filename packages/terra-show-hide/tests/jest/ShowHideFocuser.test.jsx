import React from 'react';
import { v4 as uuidv4 } from 'uuid';
import ShowHideFocuser from '../../src/ShowHideFocuser';

describe('ShowHideFocuser', () => {
  let mockSpyUuid;
  beforeAll(() => {
    mockSpyUuid = jest.spyOn(uuidv4, 'v4').mockReturnValue('00000000-0000-0000-0000-000000000000');
  });

  afterAll(() => {
    mockSpyUuid.mockRestore();
  });

  const mockPrefix = 'Patients are requesting greater affordability and efficiency in healthcare.';
  const mockFocusableText = 'In fact, the U.S. ambulatory surgery center market is expected to see a 6.9 % compound annual growth rate, reaching $33 billion by 2028. (2) Cerner understands the urgency to grow in the ambulatory surgery center market while continuing to deliver excellent care.';
  // eslint-disable-next-line react/prop-types
  const MockComponent = ({ prefix, focusableText }) => {
    const ref = React.useRef(null);
    return <p><ShowHideFocuser ref={ref} prefix={prefix} focusableText={focusableText} /></p>;
  };

  // Snapshot Tests
  it('should render divider spans and a span to focus to correctly', () => {
    const paragraph = enzyme.render(<MockComponent focusableText={mockFocusableText} />);
    expect(paragraph).toMatchSnapshot();
  });

  it('should render preview text correctly', () => {
    const paragraph = enzyme.render(<MockComponent prefix={mockPrefix} focusableText={mockFocusableText} />);
    expect(paragraph).toMatchSnapshot();
  });

  it('should render focusable text correctly', () => {
    const paragraph = enzyme.render(<MockComponent focusableText={mockFocusableText} />);
    expect(paragraph).toMatchSnapshot();
  });

  it('should throw error for required focusableText prop', () => {
    try {
      enzyme.shallow(<MockComponent />);
    } catch (e) {
      expect(e.message).toContain('The prop `focusableText` is marked as required');
    }
  });
});
