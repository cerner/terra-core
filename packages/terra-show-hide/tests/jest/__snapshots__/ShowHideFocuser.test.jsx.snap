// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowHideFocuser should render divider spans and a span to focus to correctly 1`] = `
<p>
  <span
    class="text-divider"
  />
  <span
    id="focus-ref-00000000-0000-0000-0000-000000000000"
    role="group"
    tabindex="-1"
  >
    In fact, the U.S. ambulatory surgery center market is expected to see a 6.9 % compound annual growth rate, reaching $33 billion by 2028. (2) Cerner
  </span>
  <span
    class="text-divider"
  />
   understands the urgency to grow in the ambulatory surgery center market while continuing to deliver excellent care.
</p>
`;

exports[`ShowHideFocuser should render focusable text correctly 1`] = `
<p>
  <span
    class="text-divider"
  />
  <span
    id="focus-ref-00000000-0000-0000-0000-000000000000"
    role="group"
    tabindex="-1"
  >
    In fact, the U.S. ambulatory surgery center market is expected to see a 6.9 % compound annual growth rate, reaching $33 billion by 2028. (2) Cerner
  </span>
  <span
    class="text-divider"
  />
   understands the urgency to grow in the ambulatory surgery center market while continuing to deliver excellent care.
</p>
`;

exports[`ShowHideFocuser should render preview text correctly 1`] = `
<p>
  Patients are requesting greater affordability and efficiency in healthcare. 
  <span
    class="text-divider"
  />
  <span
    id="focus-ref-00000000-0000-0000-0000-000000000000"
    role="group"
    tabindex="-1"
  >
    In fact, the U.S. ambulatory surgery center market is expected to see a 6.9 % compound annual growth rate, reaching $33 billion by 2028. (2) Cerner
  </span>
  <span
    class="text-divider"
  />
   understands the urgency to grow in the ambulatory surgery center market while continuing to deliver excellent care.
</p>
`;
